File: java/cousinsInBinaryTree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    private int x;
    private int y;
    private int k = -1;
    private int parent = -1;
    private boolean cousins = false;
    
    public boolean isCousins(TreeNode root, int x, int y) {
        this.x = x;
        this.y = y;
        visit(root, 0, 0);
        return cousins;
    }
    
    private void visit(TreeNode node, int level, int parentValue) {
        //check if this is a target node
        if (node.val == x || node.val == y) {
            //if the parent is zero, this is the root and the result is false
            if (parentValue == 0) {
                cousins = false;
                return;
            }
            //check if we've found the other target node
            if (k > 0) {
                //if so, determine whether the two nodes are cousins
                cousins = level == k && parentValue != parent;
                return;
            } else {
                //otherwise, assign k and parent to the appropriate values of this node, to check against
                //the second node when it's found
                k = level;
                parent = parentValue;
            }
        }
        //visit the other nodes in the tree recursively
        if (node.right != null) {
            visit(node.right, level + 1, node.val);
        }
        if (node.left != null) {
            visit(node.left, level + 1, node.val);
        }
    }
}

File: python/cousins_in_binary_tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isCousins(self, root: TreeNode, x: int, y: int) -> bool:
        #set k (level of cousins) and parent (value of parent of first found element) to None
        self.k = None
        self.parent = None
        #set result to false
        self.is_cousins = False
        #visit each node recursively
        def visit(level: int, parent_value: int, node: TreeNode) -> None:
            if node is None: 
                return
            #if the node is one of the target nodes, determine if it's the first or second one that's found
            if node.val in [x, y]:
                #if it's the root, the result is false
                if parent_value is None:
                    self.is_cousins = False
                    return  
                #if we've found the other node, determine if the two are cousins
                if self.k is not None:
                    self.is_cousins = self.k == level and parent_value != self.parent
                    return
                #otherwise, assign k and parent to the appropriate values
                self.k = level
                self.parent = parent_value
            if node.left is not None:
                visit(level + 1, node.val, node.left)
            if node.right is not None:
                visit(level + 1, node.val, node.right)
        visit(0, None, root)
        return self.is_cousins